src/ucan-cloud/client.ts

```diff
@@ -0,0 +1,253 @@
+  event: Link;
+  server: Principal;
+  service: ConnectionView<Service>;
+}) {
+  const invocation = ClockCaps.advance.invoke({
+    issuer: agent.signer,
+    audience: server,
+    with: clock.did(),
+    nb: { event },
+    proofs: [agent.delegation, agent.attestation],
+  });
+
+  return await invocation.execute(service);
+}
+
+/**
+ * Create a clock.
+ * Audience is always a `did:mailto` DID.
+ */
+export async function createClock({ audience }: { audience: Principal }): Promise<Clock> {
+  const signer = await ed25519.Signer.generate();
+  const delegation = await ClockCaps.clock.delegate({
+    issuer: signer,
+    audience,
+    with: signer.did(),
+    expiration: Infinity,
+  });
+
+  return {
+    delegation,
+    did: () => signer.did(),
+    signer: () => signer,
+  };
+}
+
+/**
+ * Create a clock event.
+ */
+export async function createClockEvent({ metadata }: { metadata: Uint8Array }) {
+  const eventData = { metadata };
+  const event = { parents: [], data: eventData };
+
+  const block = await Block.encode({
+    value: event,
+    codec: CBOR,
+    hasher: sha256,
+  });
+
+  return block;
+}
+
+export async function getClockHead({
+  agent,
+  clock,
+  server,
+  service,
+}: {
+  agent: Agent;
+  clock: Clock;
+  server: Principal;
+  service: ConnectionView<Service>;
+}) {
+  const invocation = ClockCaps.head.invoke({
+    issuer: agent.signer,
+    audience: server,
+    with: clock.did(),
+    proofs: [agent.delegation, agent.attestation],
+  });
+
+  return await invocation.execute(service);
+}
+
+export async function metadataFromClockEvent(eventBytes: Uint8Array): Promise<Uint8Array> {
+  const block = await Block.decode<{ data: { metadata: Uint8Array } }, 113, 18>({
+    bytes: eventBytes,
+    codec: CBOR,
+    hasher: sha256,
+  });
+
+  return block.value.data.metadata;
+}
+
+/**
+ * Register a clock.
+ */
+export async function registerClock({
+  clock,
+  server,
+  service,
+}: {
+  clock: Clock;
+  server: Principal;
+  service: ConnectionView<Service>;
+}) {
+  const invocation = ClockCaps.register.invoke({
+    issuer: clock.signer(),
+    audience: server,
+    with: clock.did(),
+    nb: { proof: clock.delegation.cid },
+    proofs: [clock.delegation],
+  });
+
+  return await invocation.execute(service);
+}
+
+////////////////////////////////////////
+// CONNECTION
+////////////////////////////////////////
+
+export const service = (server: { id: Principal; host: URI }) => {
+  const url = server.host.toString();
+
+  const channel: Channel<Service> = {
+    async request({ headers, body }) {
+      const response = await fetch(url, {
+        headers,
+        body,
+        method: "POST",
+      });
+
+      if (!response.ok) throw new Error(`HTTP Request failed. ${"POST"} ${url} → ${response.status}`);
+      const buffer = response.ok ? await response.arrayBuffer() : new Uint8Array();
+
+      return {
+        headers: response.headers.entries ? Object.fromEntries(response.headers.entries()) : {},
```

Comment:
why not stick with headers typ?

Reply:
Needed a normal object instead of an iterator.